pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -404,7 +410,16 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
 	// This also detects close.
 	var receiveError error
 	reqChannel := make(chan *xdsapi.DiscoveryRequest, 1)
-	go receiveThread(con, reqChannel, &receiveError)
+	controlChannel := make(chan int, 1)
+	defer func() {
+		// tell receive thread to quit
+		// and remove req from channel make sure receive thread can stop
+		controlChannel <- 1
+		if len(reqChannel) > 0 {
+			<-reqChannel
+		}
+	}()
+	go receiveThread(con, reqChannel, &receiveError, controlChannel)
```

Comment:
Instead of passing another `controlChannel` why not simply close the `reqChannel` and catch this on the other side?

Reply:
emm...i'm a fresher to golang . i will try it. 
thanks for reply

Reply:
yes , i have try code like this. It is simply but not easy to understand the mean of panic recover. maybe using a control channel can be more readable ？ 
```go
func receiveThread(con *XdsConnection, reqChannel chan *xdsapi.DiscoveryRequest, errP *error) {
	defer close(reqChannel) // indicates close of the remote side.
	defer func() {
		if r := recover(); r != nil {
			err, ok := r.(error)
			if ok {
				errP = &err
			}
		}
	}()
	for {
		req, err := con.stream.Recv()
		if err != nil {
			if status.Code(err) == codes.Canceled || err == io.EOF {
				con.mu.RLock()
				adsLog.Infof("ADS: %q %s terminated %v", con.PeerAddr, con.ConID, err)
				con.mu.RUnlock()
				return
			}
			*errP = err
			adsLog.Errorf("ADS: %q %s terminated with errors %v", con.PeerAddr, con.ConID, err)
			totalXDSInternalErrors.Add(1)
			return
		}
		reqChannel <- req
	}
}
```

Reply:
There's already con.doneChannel that's closed when StreamAggregatedResources returns, seems you could use that in receiveThread. Seems simpler and cleaner than abusing panic recovery.

Reply:
greet . use doneChannel is awesome!